### PR TITLE
Add Pathfinder class with basic A* algorithm

### PR DESCRIPTION
**Pull request by dominic**

### CODE REVIEW

**Comment by dominic** [comment]
pilots

File: src/map/PathFinder.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      }
+
+      // Get neighbors we haven't visited, and sort nodes to
+      // search by distance from the target destination
+      toVisit = toVisit
+        .concat(this.getNeighbors(checkCoord, to))
+        .sort((a, b) => {
+          const nodeA = this.nodes[vecToPos(a)]
+          const nodeB = this.nodes[vecToPos(b)]
+          return nodeA!.distance - nodeB!.distance
```
Comment (jeffomatic) [inline src/map/PathFinder.ts:81]: This is an interesting cost function! I think in terms of this implementation, here's how the canonical algorithms would score it:

- Djikstra: sort by `node.score`
- A*: sort by `node.score + k * node.distance`, where `k` is some tunable bias

I would be curious to see how the cost function in this PR would look if we visualized it.

File: src/map/PathFinder.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    return []
+  }
+
+  // Add nodes with distances for NSEW iff they're not a wall
+  getNeighbors(coord: [number, number], to: vec2): [number, number][] {
+    const [x, y] = coord
+    const neighborList: [number, number][] = []
+
+    const cur = this.nodes[vecToPos(coord)]
+    ;[
```
Comment (jeffomatic) [inline src/map/PathFinder.ts:94]: Semicolon?

File: src/map/PathFinder.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      [x, y - 1], // North
+      [x, y + 1], // South
+      [x - 1, y], // West
+      [x + 1, y], // East
+      // TODO - add diagonals?
```
Comment (jeffomatic) [inline src/map/PathFinder.ts:99]: Diagonals would be cool! The way to do that is to insert the diagonals with a score of `cur.score + Math.SQRT2`.

File: src/map/PathFinder.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    ].forEach(([nx, ny]) => {
+      // t2a function from editor
+      const p = vec2.sub(vec2.create(), [nx, ny], this.map.origin)
+      if (
+        this.map.entities[p[1] * this.map.dimensions[0] + p[0]] != Type.WALL
```
Comment (jeffomatic) [inline src/map/PathFinder.ts:104]: What would be cool is to have a "current walls" construct hanging around in `Game` that we could update when wall entities are destroyed, so the pathfinder can dynamically update when walls get destroyed.

File: src/systems/pilot.ts [hunk 5]
```diff
@@ -3,13 +3,33 @@ import { glMatrix, vec2 } from 'gl-matrix'
...
+      ),
+    )
+    const tilePos = vec2.floor(
+      vec2.create(),
+      vec2.scale(vec2.create(), destPos, 1 / TILE_SIZE),
+    )
+
+    // Generate path to tile position
+    const pathfinder = new PathFinder(g.map)
+    const path = pathfinder.discover(playerTilePos, tilePos)
```
Comment (jeffomatic) [inline src/systems/pilot.ts:30]: Pathfinding during spawning is a good optimization. I have been wondering whether we eventually have to worry about pathfinding every frame (or every few frames) if the target moves (e.g., the pilot returning to the tank), or obstacles update dynamically (e.g., walls are destroyed). Maybe there is a middle ground where we keep a list of all active paths, and re-path them on certain game events.

File: src/map/PathFinder.ts (hunk 2 above)
Comment (jeffomatic) [reply src/map/PathFinder.ts:94]: Oh I think I get it. That's....wild.

**Pull request merged**